File: solutions/javascript/class-design/minStack.js
function getLastElement(array) {
    const lastIdx = array.length - 1;
    return array[lastIdx];
}

class MinStack {
    constructor() {
        this.stack = [];
        this.minStack = [];
    }

    push(element) {
        this.stack.push(element);

        const isNewMin = this.minStack.length === 0 || element <= this.getMin();
        if (isNewMin) {
            this.minStack.push(element);
        }
    }

    pop() {
        if (getLastElement(this.minStack) === getLastElement(this.stack)) {
            this.minStack.pop();
        }
        this.stack.pop();
    }

    top() {
        return getLastElement(this.stack);
    }

    getMin() {
        return getLastElement(this.minStack);
    }
}

File: solutions/python/class-design/min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, element):
        self.stack.append(element)

        isNewMin = len(self.min_stack) == 0 or element <= self.getMin()
        if isNewMin:
            self.min_stack.append(element)

    def pop(self):
        if self.min_stack[-1] == self.stack[-1]:
            self.min_stack.pop()
        self.stack.pop()

    def top(self):
        return self.stack[-1]

    def getMin(self):
        return self.min_stack[-1]
